File: transports/ocaml/test/max_frame.ml
(* The configurable max-frame guard (conventions doc section 5): a host sets the
   limit up or down through the carrier's public API, the limit applies to reads and
   writes alike, an oversized inbound length is rejected before allocation, and an
   invalid limit fails at construction rather than on the first frame.

   The framing functions take channels, so each case drives a real temp-file channel
   pair rather than a mock: the guard is proven against the same I/O path a socket
   carrier uses. *)

open Csilgen_transport

let check name cond = if not cond then failwith ("max_frame: " ^ name)

(* Run [f] with an out_channel, then reopen the same file as an in_channel so a
   written frame can be read back through the real framing path. *)
let with_roundtrip_channels f =
  let path = Filename.temp_file "csil_max_frame" ".bin" in
  let oc = open_out_bin path in
  let result = f oc path in
  close_out oc;
  Sys.remove path;
  result

let is_frame_too_large = function
  | Error (Conventions.Frame_too_large _) -> true
  | _ -> false

let is_invalid_max_frame = function
  | Error (Conventions.Invalid_max_frame _) -> true
  | _ -> false

(* Write [body] with the given guard, then read it back with the same guard. *)
let write_then_read ?max_frame body =
  let path = Filename.temp_file "csil_max_frame" ".bin" in
  let oc = open_out_bin path in
  let max = match max_frame with Some m -> m | None -> Conventions.max_frame_default in
  let wrote = Carrier.write_length_prefixed oc body ~max in
  close_out oc;
  let read =
    match wrote with
    | Error _ -> Ok None
    | Ok () ->
        let ic = open_in_bin path in
        let r = Carrier.read_length_prefixed ic ~max in
        close_in ic;
        r
  in
  Sys.remove path;
  (wrote, read)

let default_accepts_below () =
  let body = Bytes.make 1024 '\xAB' in
  let wrote, read = write_then_read body in
  check "default accepts a frame below it" (wrote = Ok ());
  check "default reads the frame back" (read = Ok (Some body))

let default_rejects_above () =
  let body = Bytes.make (Conventions.max_frame_default + 1) '\x00' in
  let wrote, _ = write_then_read body in
  check "default rejects a frame above it" (is_frame_too_large wrote)

let larger_limit_accepts () =
  let body = Bytes.make (Conventions.max_frame_default + 1) '\x00' in
  let raised = Conventions.max_frame_default + 4096 in
  let wrote, read = write_then_read ~max_frame:raised body in
  check "raised limit accepts the frame" (wrote = Ok ());
  check "raised limit reads the frame back" (read = Ok (Some body))

let smaller_limit_rejects () =
  let body = Bytes.make 1024 '\xCD' in
  let wrote, _ = write_then_read ~max_frame:64 body in
  check "lowered limit rejects the frame" (is_frame_too_large wrote)

let oversized_length_rejected_before_allocation () =
  (* A prefix claiming ~4 GiB followed by no body: if the guard ran after the read
     this would allocate; it must fail on the prefix alone. *)
  let path = Filename.temp_file "csil_max_frame" ".bin" in
  let oc = open_out_bin path in
  output_bytes oc (Bytes.of_string "\xFF\xFF\xFF\xFF");
  close_out oc;
  let ic = open_in_bin path in
  let read = Carrier.read_length_prefixed ic ~max:4096 in
  close_in ic;
  Sys.remove path;
  check "oversized inbound length rejected" (is_frame_too_large read)

let invalid_limits_rejected () =
  List.iter
    (fun limit ->
      check
        (Printf.sprintf "limit %d must be rejected" limit)
        (is_invalid_max_frame (Conventions.validate_max_frame limit)))
    [ 0; -1; -4096; Conventions.max_frame_limit + 1 ];
  (* And at construction, through the carrier the host actually builds. *)
  with_roundtrip_channels (fun oc path ->
      let ic = open_in_bin path in
      let built =
        Carrier.stream_carrier_with_max_frame ~max_frame:0 ic oc
      in
      close_in ic;
      check "carrier construction rejects an invalid limit"
        (match built with
        | Error (Conventions.Invalid_max_frame _) -> true
        | _ -> false))

let boundary_limits_accepted () =
  List.iter
    (fun limit ->
      check
        (Printf.sprintf "limit %d must be accepted" limit)
        (Conventions.validate_max_frame limit = Ok limit))
    [ 1; Conventions.max_frame_default; Conventions.max_frame_limit ]

let run () =
  default_accepts_below ();
  default_rejects_above ();
  larger_limit_accepts ();
  smaller_limit_rejects ();
  oversized_length_rejected_before_allocation ();
  invalid_limits_rejected ();
  boundary_limits_accepted ();
  print_endline "max_frame: all guard checks passed"

File: transports/rust/tests/max_frame.rs
//! The configurable max-frame guard (conventions doc §5): a host sets the limit up
//! or down through the carrier's public API, the limit applies to reads and writes
//! alike, an oversized inbound length is rejected before allocation, and an invalid
//! limit fails at construction rather than on the first frame.

use csilgen_transport::carrier::{FrameCarrier, StreamCarrier};
use csilgen_transport::conventions::{MAX_FRAME_DEFAULT, MAX_FRAME_LIMIT, TransportError};
use std::io::{Cursor, Read, Write};

/// A stream whose reads are counted, so a test can prove the guard fires before the
/// frame body is ever pulled off the wire.
struct CountingStream {
    inner: Cursor<Vec<u8>>,
    read: usize,
}

impl CountingStream {
    fn new(bytes: Vec<u8>) -> Self {
        Self {
            inner: Cursor::new(bytes),
            read: 0,
        }
    }
}

impl Read for CountingStream {
    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
        let n = self.inner.read(buf)?;
        self.read += n;
        Ok(n)
    }
}

impl Write for CountingStream {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        Ok(buf.len())
    }
    fn flush(&mut self) -> std::io::Result<()> {
        Ok(())
    }
}

/// An in-memory duplex: writes land in a buffer that subsequent reads drain.
#[derive(Default)]
struct Duplex {
    buf: Vec<u8>,
    pos: usize,
}

impl Read for Duplex {
    fn read(&mut self, out: &mut [u8]) -> std::io::Result<usize> {
        let n = (&self.buf[self.pos..]).read(out)?;
        self.pos += n;
        Ok(n)
    }
}

impl Write for Duplex {
    fn write(&mut self, data: &[u8]) -> std::io::Result<usize> {
        self.buf.extend_from_slice(data);
        Ok(data.len())
    }
    fn flush(&mut self) -> std::io::Result<()> {
        Ok(())
    }
}

#[test]
fn default_limit_accepts_frame_below_it() {
    let mut carrier = StreamCarrier::new(Duplex::default());
    let frame = vec![0xAB; 1024];
    carrier.send_frame(&frame).expect("frame under the limit");
    let got = carrier.recv_frame().expect("read back").expect("a frame");
    assert_eq!(got, frame);
}

#[test]
fn default_limit_rejects_frame_above_it() {
    let mut carrier = StreamCarrier::new(Duplex::default());
    let frame = vec![0u8; MAX_FRAME_DEFAULT + 1];
    match carrier.send_frame(&frame) {
        Err(TransportError::FrameTooLarge { got, max }) => {
            assert_eq!(got, MAX_FRAME_DEFAULT + 1);
            assert_eq!(max, MAX_FRAME_DEFAULT);
        }
        other => panic!("expected FrameTooLarge, got {other:?}"),
    }
}

#[test]
fn larger_custom_limit_accepts_what_default_rejects() {
    let mut carrier = StreamCarrier::with_max_frame(Duplex::default(), MAX_FRAME_DEFAULT + 4096)
        .expect("a limit above the default is valid");
    let frame = vec![0u8; MAX_FRAME_DEFAULT + 1];
    carrier.send_frame(&frame).expect("raised limit accepts");
    let got = carrier.recv_frame().expect("read back").expect("a frame");
    assert_eq!(got.len(), frame.len());
}

#[test]
fn smaller_custom_limit_rejects_what_default_accepts() {
    let mut carrier =
        StreamCarrier::with_max_frame(Duplex::default(), 64).expect("64 is a valid limit");
    let frame = vec![0xCD; 1024];
    assert!(matches!(
        carrier.send_frame(&frame),
        Err(TransportError::FrameTooLarge { got: 1024, max: 64 })
    ));
}

#[test]
fn oversized_incoming_length_rejected_before_allocation() {
    // A prefix claiming ~4 GiB followed by no body: if the guard ran after the read
    // this would allocate; it must fail on the prefix alone.
    let stream = CountingStream::new(vec![0xFF, 0xFF, 0xFF, 0xFF]);
    let mut carrier = StreamCarrier::with_max_frame(stream, 4096).expect("4096 is valid");
    assert!(matches!(
        carrier.recv_frame(),
        Err(TransportError::FrameTooLarge { .. })
    ));
    assert_eq!(
        carrier.into_inner().read,
        4,
        "guard must fire on the 4-byte prefix alone"
    );
}

#[test]
fn invalid_limits_rejected_at_construction() {
    for limit in [0usize, MAX_FRAME_LIMIT + 1, usize::MAX] {
        match StreamCarrier::with_max_frame(Duplex::default(), limit) {
            Err(TransportError::InvalidMaxFrame { got, limit: lim }) => {
                assert_eq!(got, limit);
                assert_eq!(lim, MAX_FRAME_LIMIT);
            }
            _ => panic!("limit {limit} must be rejected"),
        }
    }
    for limit in [1usize, MAX_FRAME_DEFAULT, MAX_FRAME_LIMIT] {
        assert!(
            StreamCarrier::with_max_frame(Duplex::default(), limit).is_ok(),
            "limit {limit} must be accepted"
        );
    }
}
